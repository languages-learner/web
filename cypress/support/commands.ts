/// <reference types="cypress" />

import { elSelector } from '@@/cypress/utils'
import { type EDataTest, type EDataTestClass } from '@/enums/EDataTest'

Cypress.Commands.add('el', (dataTest: EDataTest) => cy.get(elSelector(dataTest)))
Cypress.Commands.add('elByClass', (dataTestClass: EDataTestClass) => cy.get(`.${dataTestClass}`))

Cypress.Commands.add('clickOutside', () => {
    cy.log('click outside')

    return cy
        .get('body', { log: false })
        .click(0,0, { log: false })
})
