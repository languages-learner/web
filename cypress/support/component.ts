/// <reference types="cypress" />

import './commands'
import { mount } from 'cypress/vue'
import { createPinia } from 'pinia'
import { createI18n } from 'vue-i18n'

import 'firebase/compat/auth'
import { TEST_INTERFACE_LANGUAGE_NAME } from '@/const/InterfaceLanguage'
import { createAppRouter } from '@/router'

// import '@/assets/main.scss'
Cypress.Commands.add('mount', (component, options = {}) => {
    options.global = options.global || {}
    options.global.plugins = options.global.plugins || []
    options.global.plugins.push(createPinia())
    options.global.plugins.push({
        async install(app) {
            const router = createAppRouter()
            await router.push({ params: { lang: TEST_INTERFACE_LANGUAGE_NAME } })
            await router.isReady()
            app.use(createI18n({
                locale: TEST_INTERFACE_LANGUAGE_NAME,
                fallbackLocale: {
                    default: [TEST_INTERFACE_LANGUAGE_NAME],
                },
                messages: {
                    [TEST_INTERFACE_LANGUAGE_NAME]: {},
                },
                legacy: false,
                globalInjection: true,
            }))
            app.use(createPinia())
        },
    })

    return mount(component, options)
})
