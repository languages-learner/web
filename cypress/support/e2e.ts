/// <reference types="cypress" />

import './commands'
import { withLang } from '@@/cypress/utils'
import { EDataTest } from '@/enums/EDataTest'

beforeEach(() => {
    cy.intercept({ resourceType: /xhr|fetch/ }, { log: false })
})

Cypress.Commands.add('authWithoutSession', (options = {}) => {
    const finalOptions: Cypress.LoginOptions = {
        ...(Cypress.env('testUser') ?? {}),
        ...{
            type: 'signin',
            visitLoginPage: true,
            validateAuth: true,
        },
        ...options,
    }
    const isSignIn = finalOptions.type === 'signin'
    // TODO Maybe replace to check path
    if (finalOptions.visitLoginPage)
        cy.visit(withLang())
    cy.el(isSignIn ? EDataTest.landing_sign_in_button : EDataTest.landing_sign_up_button).click()
    if (finalOptions.username)
        cy.el(EDataTest.authentication_modal_email).type(finalOptions.username)
    if (finalOptions.password)
        cy.el(EDataTest.authentication_modal_password).type(finalOptions.password)
    cy.el(EDataTest.authentication_modal_action_button)
        // .contains(isSignIn ? 'sign_in' : 'sign_up')
        .click()
    if (finalOptions.validateAuth)
        cy.location('pathname').should('equal', withLang('/dictionary'))
    else
        cy.wait(1000)
})

Cypress.Commands.add('auth', (options = {}) => {
    cy.session(
        [options.username, options.password, options.type],
        () => {
            cy.logout()
            cy.authWithoutSession(options)
        },
        {
            validate: async () => {
                cy.wrap(null).then(async () => {
                    expect((await indexedDB.databases()).map<string>(db => db.name ?? '').indexOf('firebaseLocalStorageDb')).not.to.equal(-1)
                })
                cy.visit(withLang('/dictionary'))
                cy.location('pathname').should('equal', withLang('/dictionary'))
                cy.waitWorkspacePageInit()
            },
        },
    )
})

Cypress.Commands.add('logout', () => {
    indexedDB.deleteDatabase('firebaseLocalStorageDb')
})

Cypress.Commands.add('waitWorkspacePageInit', () => {
    cy.el(EDataTest.workspace_header_user_avatar, { timeout: 15000 }).should('exist').and('not.be.empty')
})
