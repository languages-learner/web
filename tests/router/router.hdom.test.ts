import { describe, expect } from 'vitest'
import { testApp } from '@@/tests/support/testApp'

describe('router', () => {
    testApp('should redirect unauthenticated user to landing page', async ({ app }) => {
        const router = app.runWithContext(() => useRouter())
        await router.push({ name: EPageName.DICTIONARY })
        expect(router.currentRoute.value.name).toBe(EPageName.LANDING)
    })
})
