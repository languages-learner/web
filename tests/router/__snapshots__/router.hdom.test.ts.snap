// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`router > redirect 1`] = `
{
  "fullPath": "/",
  "hash": "",
  "href": "/",
  "matched": [
    {
      "aliasOf": undefined,
      "beforeEnter": undefined,
      "children": [],
      "components": undefined,
      "enterCallbacks": {},
      "instances": {},
      "leaveGuards": Set {},
      "meta": {},
      "name": undefined,
      "path": "",
      "props": {},
      "redirect": [Function],
      "updateGuards": Set {},
    },
  ],
  "meta": {},
  "name": undefined,
  "params": {},
  "path": "/",
  "query": {},
  "redirectedFrom": undefined,
}
`;
